src/libraries/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFileExtensions.ZipArchive.Create.Async.cs

```diff
@@ -0,0 +1,125 @@
+    /// <exception cref="IOException">An I/O error occurred while opening the file specified by sourceFileName.</exception>
+    /// <exception cref="UnauthorizedAccessException">sourceFileName specified a directory.
+    /// -or- The caller does not have the required permission.</exception>
+    /// <exception cref="FileNotFoundException">The file specified in sourceFileName was not found. </exception>
+    /// <exception cref="NotSupportedException">sourceFileName is in an invalid format or the ZipArchive does not support writing.</exception>
+    /// <exception cref="ObjectDisposedException">The ZipArchive has already been closed.</exception>
+    /// <exception cref="OperationCanceledException">An asynchronous operation is cancelled.</exception>
+    ///
+    /// <param name="destination">The zip archive to add the file to.</param>
+    /// <param name="sourceFileName">The path to the file on the file system to be copied from. The path is permitted to specify relative
+    /// or absolute path information. Relative path information is interpreted as relative to the current working directory.</param>
+    /// <param name="entryName">The name of the entry to be created.</param>
+    /// <param name="compressionLevel">The level of the compression (speed/memory vs. compressed size trade-off).</param>
+    /// <param name="cancellationToken">The cancellation token to monitor for cancellation requests.</param>
+    /// <returns>A wrapper for the newly created entry.</returns>
+    public static Task<ZipArchiveEntry> CreateEntryFromFileAsync(this ZipArchive destination,
+                                                      string sourceFileName, string entryName, CompressionLevel compressionLevel, CancellationToken cancellationToken = default) =>
+        DoCreateEntryFromFileAsync(destination, sourceFileName, entryName, compressionLevel, cancellationToken);
+
+    internal static async Task<ZipArchiveEntry> DoCreateEntryFromFileAsync(this ZipArchive destination,
+                                                          string sourceFileName, string entryName, CompressionLevel? compressionLevel, CancellationToken cancellationToken = default)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        ArgumentNullException.ThrowIfNull(destination);
+        ArgumentNullException.ThrowIfNull(sourceFileName);
+        ArgumentNullException.ThrowIfNull(entryName);
+
+        // Checking of compressionLevel is passed down to DeflateStream and the IDeflater implementation
+        // as it is a pluggable component that completely encapsulates the meaning of compressionLevel.
+
+        // Argument checking gets passed down to FileStream's ctor and CreateEntry
+
+        FileStream fs = new FileStream(sourceFileName, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 0x1000, useAsync: true);
```

Comment:
Let's not duplicate the buffer size constant everywhere.

Reply:
If we decide to change it, we can do it in a separate PR.

Reply:
> If we decide to change it, we can do it in a separate PR.

Yes, but how about making it an argument that can be provided by the users? It can greatly improve perf, so I would expect it to be configurable.

Reply:
Removed the code duplication in https://github.com/dotnet/runtime/pull/114421/commits/85af5798547c75f412dcc4759068f2b8635ff652

Increased bufferSize to 16K in https://github.com/dotnet/runtime/pull/114421/commits/1b9770d9aabfecf035d28a2b6c9d79279a893511